services/vision/obstaclesdepth/obstacles_depth.go

```diff
@@ -0,0 +1,323 @@
+	sinTheta    float64
+	intrinsics  *transform.PinholeCameraIntrinsics
+	returnPCDs  bool
+	k           int
+	depthStream gostream.VideoStream
+}
+
+const (
+	// the first 3 consts are parameters from Manduchi et al.
+	defaultHmin     = 0.0
+	defaultHmax     = 1.0
+	defaultThetamax = 45
+
+	defaultK = 10 // default number of obstacle segments to create
+	sampleN  = 4  // we sample 1 in every sampleN depth points
+)
+
+func init() {
+	resource.RegisterService(svision.API, model, resource.Registration[svision.Service, *ObsDepthConfig]{
+		DeprecatedRobotConstructor: func(ctx context.Context, r any, c resource.Config, logger golog.Logger) (svision.Service, error) {
+			attrs, err := resource.NativeConfig[*ObsDepthConfig](c)
+			if err != nil {
+				return nil, err
+			}
+			actualR, err := utils.AssertType[robot.Robot](r)
+			if err != nil {
+				return nil, err
+			}
+			return registerObstaclesDepth(ctx, c.ResourceName(), attrs, actualR, logger)
+		},
+	})
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *ObsDepthConfig) Validate(path string) ([]string, error) {
+	deps := []string{}
+	if config.Hmin >= config.Hmax {
+		return nil, errors.New("Hmin should be less than Hmax")
+	}
+	if config.Hmin < 0 {
+		return nil, errors.New("Hmin should be greater than or equal to 0")
+	}
+	if config.Hmax < 0 {
+		return nil, errors.New("Hmax should be greater than or equal to 0")
+	}
+	if config.ThetaMax < 0 || config.ThetaMax > 360 {
+		return nil, errors.New("ThetaMax should be in degrees between 0 and 360")
+	}
+
+	return deps, nil
+}
+
+func registerObstaclesDepth(
+	ctx context.Context,
+	name resource.Name,
+	conf *ObsDepthConfig,
+	r robot.Robot,
+	logger golog.Logger,
+) (svision.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::registerObstacleDistanceDetector")
+	defer span.End()
+	if conf == nil {
+		return nil, errors.New("config for obstacles_depth cannot be nil")
+	}
+
+	// Use defaults if needed
+	if conf.Hmax == 0 {
+		conf.Hmax = defaultHmax
+	}
+	if conf.ThetaMax == 0 {
+		conf.ThetaMax = defaultThetamax
+	}
+
+	sinTheta := math.Sin(conf.ThetaMax * math.Pi / 180) // sin(radians(theta))
+	myObsDep := obsDepth{
+		hMin: 1000 * conf.Hmin, hMax: 1000 * conf.Hmax, sinTheta: sinTheta,
+		returnPCDs: conf.ReturnPCDs, k: defaultK,
+	}
+
+	segmenter := myObsDep.buildObsDepth(logger) // does the thing
+	return svision.NewService(name, r, nil, nil, nil, segmenter)
+}
+
+// BuildObsDepth will check for intrinsics and determine how to build based on that.
+func (o *obsDepth) buildObsDepth(logger golog.Logger) func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		props, err := src.Properties(ctx)
+		if err != nil || props.IntrinsicParams == nil {
+			logger.Warn("obstacles depth started without camera's intrinsic parameters")
+			ni := o.buildObsDepthNoIntrinsics()
+			return ni(ctx, src)
```

Comment:
You don't have to split this into two lines. make a method called 
```
func (o *obsDepth) obsDepthNoIntrinsics(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) { ... }
```
and then just call it:
`return o.obsDepthNoIntrinsics(ctx, src)`